File: src/rbs_rag/text.py
from __future__ import annotations

import math
import re
from collections import Counter
from collections.abc import Iterable

TOKEN_RE = re.compile(r"[A-Za-z0-9][A-Za-z0-9_'-]*")


def tokenize(text: str) -> list[str]:
    return [match.group(0).lower() for match in TOKEN_RE.finditer(text)]


def cosine_similarity(left: list[float], right: list[float]) -> float:
    if not left or not right or len(left) != len(right):
        return 0.0
    return sum(a * b for a, b in zip(left, right))


def l2_normalize(values: Iterable[float]) -> list[float]:
    vector = list(values)
    norm = math.sqrt(sum(value * value for value in vector))
    if norm == 0:
        return vector
    return [value / norm for value in vector]


def bm25_scores(query: str, documents: dict[str, str], k1: float = 1.5, b: float = 0.75) -> dict[str, float]:
    query_terms = tokenize(query)
    if not query_terms or not documents:
        return {doc_id: 0.0 for doc_id in documents}

    tokenized = {doc_id: tokenize(text) for doc_id, text in documents.items()}
    lengths = {doc_id: len(tokens) for doc_id, tokens in tokenized.items()}
    avg_length = sum(lengths.values()) / max(len(lengths), 1)
    doc_freq: Counter[str] = Counter()
    for tokens in tokenized.values():
        doc_freq.update(set(tokens))

    total_docs = len(documents)
    scores: dict[str, float] = {}
    for doc_id, tokens in tokenized.items():
        frequencies = Counter(tokens)
        score = 0.0
        doc_length = max(lengths[doc_id], 1)
        for term in query_terms:
            if frequencies[term] == 0:
                continue
            idf = math.log(1 + (total_docs - doc_freq[term] + 0.5) / (doc_freq[term] + 0.5))
            numerator = frequencies[term] * (k1 + 1)
            denominator = frequencies[term] + k1 * (1 - b + b * doc_length / max(avg_length, 1))
            score += idf * numerator / denominator
        scores[doc_id] = score
    return scores

File: rbs-rag-node/src/text.ts
const TOKEN_RE = /[A-Za-z0-9][A-Za-z0-9_'-]*/g;

export function tokenize(text: string): string[] {
  const matches: string[] = [];
  let match: RegExpExecArray | null;
  TOKEN_RE.lastIndex = 0;
  while ((match = TOKEN_RE.exec(text)) !== null) {
    matches.push(match[0].toLowerCase());
  }
  return matches;
}

export function cosineSimilarity(left: number[], right: number[]): number {
  if (!left.length || !right.length || left.length !== right.length) return 0;
  let dot = 0, normLeft = 0, normRight = 0;
  for (let i = 0; i < left.length; i++) {
    dot += left[i] * right[i];
    normLeft += left[i] * left[i];
    normRight += right[i] * right[i];
  }
  normLeft = Math.sqrt(normLeft);
  normRight = Math.sqrt(normRight);
  if (normLeft === 0 || normRight === 0) return 0;
  return dot / (normLeft * normRight);
}

export function l2Normalize(values: number[]): number[] {
  let norm = 0;
  for (const v of values) norm += v * v;
  norm = Math.sqrt(norm);
  if (norm === 0) return values;
  return values.map(v => v / norm);
}

export function bm25Scores(
  query: string,
  documents: Record<string, string>,
  k1 = 1.5,
  b = 0.75
): Record<string, number> {
  const queryTerms = tokenize(query);
  const docIds = Object.keys(documents);
  if (!queryTerms.length || !docIds.length) {
    const result: Record<string, number> = {};
    for (const id of docIds) result[id] = 0;
    return result;
  }

  const tokenized: Record<string, string[]> = {};
  const lengths: Record<string, number> = {};
  for (const id of docIds) {
    tokenized[id] = tokenize(documents[id]);
    lengths[id] = tokenized[id].length;
  }
  const avgLength = docIds.reduce((s, id) => s + lengths[id], 0) / Math.max(docIds.length, 1);

  const docFreq: Record<string, number> = {};
  for (const id of docIds) {
    const seen = new Set(tokenized[id]);
    for (const term of seen) {
      docFreq[term] = (docFreq[term] || 0) + 1;
    }
  }

  const totalDocs = docIds.length;
  const scores: Record<string, number> = {};
  for (const id of docIds) {
    const freqs: Record<string, number> = {};
    for (const t of tokenized[id]) freqs[t] = (freqs[t] || 0) + 1;

    let score = 0;
    const docLength = Math.max(lengths[id], 1);
    for (const term of queryTerms) {
      const f = freqs[term] || 0;
      if (f === 0) continue;
      const idf = Math.log(1 + (totalDocs - (docFreq[term] || 0) + 0.5) / ((docFreq[term] || 0) + 0.5));
      const numerator = f * (k1 + 1);
      const denominator = f + k1 * (1 - b + b * docLength / Math.max(avgLength, 1));
      score += idf * numerator / denominator;
    }
    scores[id] = score;
  }
  return scores;
}
